Feature/gp gender

markysoft opened the pull request:
Added GP counts and names to the output data.
Improved test coverage, split some mappers into own files.
Updated the info logged when retrieving data to give an indication of the progress.
As the process now takes two hours due to rate limiting I have begun making the code reentrant.
So far this is only for the ID list building part of the process.

st3v3nhunt commented on app/lib/fsHelper.js:
```diff
@@ -0,0 +1,18 @@
+const fs = require('fs');
+const log = require('./logger');
+
+function saveJson(obj, filename) {
```
Should the function be named `saveJsonSync`?

st3v3nhunt commented on app/lib/fsHelper.js:
```diff
@@ -0,0 +1,18 @@
+  log.info(`${filename} saved`);
+}
+
+function loadJson(path) {
```
Rename function to be `loadJsonSync`?

st3v3nhunt commented on app/lib/xmlJsonHelper.js:
```diff
@@ -0,0 +1,13 @@
+// the XML2Json parser cannot distinguish between a single tag and what is
+// usually a list of tags but happens to only have one item this makes sure
+// an array is an array
+function asArray(value) {
```
This function seems more generic than being in a file called `xmlJsonHelper`. It might be easier for others to find and therefore reuse if the file was called something like `utils` (which isn't a great name but is something I would have a look in to find this type of function).

st3v3nhunt commented on app/lib/mappers/mapGpCounts.js:
```diff
@@ -0,0 +1,30 @@
+}
+
+function emptyObjectToUndefined(gpCounts) {
+  return Object.keys(gpCounts).length > 0 ? gpCounts : undefined;
```
This looks like a good candidate to go in a `utils` module although with only being used once it might be a bit  early to do that...

st3v3nhunt commented on app/lib/mappers/mapOpeningTimes.js:
```diff
@@ -46,11 +52,30 @@ function mapDaysOfWeek(rawTimes) {
 function mapOpeningTimes(rawTimes) {
   return openingTimeValid(rawTimes) ? mapDaysOfWeek(rawTimes) : undefined;
 }
+function sessionTypesValid(openingTimes) {
```
Additional line spacing please

st3v3nhunt commented on test/unit/mapDoctors.js:
```diff
@@ -0,0 +1,53 @@
+  it('should map array of doctors', () => {
+    const rawDoctors = {
+      's:doctor': [
+        'Dr Manhattan',
```
👍 
👍

st3v3nhunt commented on test/unit/mapOverview.js:
```diff
@@ -17,28 +18,21 @@ describe('overview mapper', () => {
     expect(overview.openingTimes.reception).to.exist;
     expect(overview.openingTimes.surgery).to.exist;
+    expect(overview.address).to.exist;
+    expect(overview.location).to.exist;
```
Repetition of line 27?

markysoft commented:
Renamed synchronous functions with sync suffix, output files written to output folder, moved useful functions to utils library, changed regexes to literal notation https://github.com/nhsuk/gp-data-etl/issues/6
